### PR TITLE
add tested IEC 62056-21 meter 1.x version to OH2

### PR DESCRIPTION
**Pull request by kreutpet**
added 1.x binding according instruction 
http://docs.openhab.org/developers/development/compatibilitylayer.html#how-to-add-a-successfully-tested-1x-add-on-to-the-distribution

it has been discussed here https://community.openhab.org/t/iec-62056-21-meter-binding/18382/11

### CODE REVIEW

**Comment by watou** [comment]
I had seen a case where a config property stored via PaperUI resulted in a ClassCastException in the binding because the object stored in the config was not a string.  So for those bindings I changed all [string casts](https://github.com/kreutpet/openhab/blob/008278050e893c38fdc844deff0c86e738b57d27/bundles/binding/org.openhab.binding.iec6205621meter/src/main/java/org/openhab/binding/iec6205621meter/internal/Iec6205621MeterBinding.java#L199)

```java
(String) config.get("xxx")
```

to `Objects.toString()` calls

```java
Objects.toString(config.get("xxx"), null)
```

That way the actual class of the object stored in the config won't matter.

**Comment by kreutpet** [comment]
concerning adding ESH-INF to the build.properties file.
i guess i need to check the ESH-INF folder to be added to binary build, right?

**Comment by kreutpet** [comment]
sorry that i did not recognized the pill request from watou.
what would be the best way to get both merged into one pull request?

Concerning the conflict on feature.xml
As i did not know how to create the pull request from eclipse I manually added the changes in github.
how to get this resolved?

**Comment by watou** [comment]
> what would be the best way to get both merged into one pull request?

They can be merged at the same time with no ill effects, so it's OK that they stay separate.

> Concerning the conflict on feature.xml

If you rebase your branch to the current master branch, you can resolve the conflicts through that process.

**Comment by 9037568** [comment]
I think a new line needs to be added to [pom.xml](https://github.com/openhab/openhab1-addons/blob/master/features/openhab-addons-external/pom.xml), too.

Is this accurate, John? (@watou)

File: bundles/binding/org.openhab.binding.iec6205621meter/ESH-INF/binding/binding.xml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        </parameter>
+        <parameter name="baudRateChangeDelay" type="integer">
+            <label>Delay of baud rate change</label>
+            <description>Delay of baud rate change in ms. Default is 0. USB to serial converters often require a delay of up to 250ms default is 0ms</description>
+            <default>0</default>
+        </parameter>        
+        <parameter name="serialPort" type="text" required="true">
+            <label>Serial port</label>
+            <description>The serial port to use for connecting to the metering device e.g. COM1 for Windows and /dev/ttyS0 or /dev/ttyUSB0 for Linux.</description>
+            <default>0</default>
```
Comment (watou) [inline bundles/binding/org.openhab.binding.iec6205621meter/ESH-INF/binding/binding.xml]: This is not a sensible default, and I don't think that there should be a default value.

File: features/openhab-addons-external/src/main/resources/conf/iec6205621meter.cfg [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#############################  IEC 620562-21 Meter Binding ############################
+#
+# the serial port to use for connecting to the metering device e.g. COM1 for Windows
+# and /dev/ttyS0 or /dev/ttyUSB0 for Linux
+#<meter-id1>.serialPort=/dev/ttyS0
+meter1.serialPort=/dev/ttyS0
```
Comment (watou) [inline features/openhab-addons-external/src/main/resources/conf/iec6205621meter.cfg]: This .cfg file should be installed with all properties commented out (with a single `#` before the property and not followed by a space.).  The point is that the file is installed as an example to follow, not an actual configuration.

File: features/openhab-addons/src/main/feature/feature.xml [hunk 3]
```diff
@@ -506,6 +506,14 @@
         <configfile finalname="${openhab.conf}/services/zwave.cfg" override="false">mvn:${project.groupId}/openhab-addons-external/${project.version}/cfg/zwave</configfile>
     </feature>
 
+    <feature name="openhab-binding-iec6205621meter" description="IEC 62056-21 meter" version="${project.version}">
+        <feature>openhab-runtime-base</feature>
+        <feature>openhab-runtime-compat1x</feature>
+	<feature>openhab-transport-serial</feature>
```
Comment (watou) [inline features/openhab-addons/src/main/feature/feature.xml]: Seems to be a problem with spaces or tabs

File: bundles/binding/org.openhab.binding.iec6205621meter/ESH-INF/binding/binding.xml (hunk 1 above)
Comment (watou) [inline bundles/binding/org.openhab.binding.iec6205621meter/ESH-INF/binding/binding.xml]: Please rebase your branch to the current master, which will bring in the new README.md file, and make sure all of README.md, binding.xml, iec6205621meter.cfg and the actual binding code are in agreement.

File: features/openhab-addons/src/main/feature/feature.xml [hunk 4]
```diff
@@ -506,6 +506,14 @@
         <configfile finalname="${openhab.conf}/services/zwave.cfg" override="false">mvn:${project.groupId}/openhab-addons-external/${project.version}/cfg/zwave</configfile>
     </feature>
 
+    <feature name="openhab-binding-iec6205621meter" description="IEC 62056-21 meter" version="${project.version}">
```
Comment (watou) [inline features/openhab-addons/src/main/feature/feature.xml]: The feature name must be `openhab-binding-iec6205621meter1`.

**Comment by watou** [comment]
Last steps that I can see:

* features/openhab-addons-external/pom.xml: move line to alphabetically correct position.
* features/openhab-addons/src/main/feature/feature.xml: name feature "openhab-binding-iec6205621meter1" and move section to alphabetically correct position in file.
* resolve the conflict with master

File: features/openhab-addons/src/main/feature/feature.xml (hunk 4 above)
Comment (kreutpet) [inline features/openhab-addons/src/main/feature/feature.xml]: No the meter1 is coming from the configuration. 
In the configuration you can add multiple meters.
When implementing the binding into a new OH2 binding I hope to find good way to allow user to configure multiple meters within PaperUI

**Comment by watou** [comment]
Please replace your features/openhab-addons/src/main/feature/feature.xml with the latest one from the master branch, and then only add your one new feature named "openhab-binding-iec6205621meter1" in the correct place in the file, do not change the line endings of the file so the diff for features/openhab-addons/src/main/feature/feature.xml only shows your specific change.  Thanks!

**Comment by kreutpet** [comment]
i tried sveral times to get control of git on my windows PC. Unfortunately.
Will try a clean restart which may require a new pull request. Sorry

**Comment by watou** [comment]
> Will try a clean restart which may require a new pull request.

A new pull request is no bother at all on my side.  Sorry for the difficulties!  Just put #5077 in a comment in your new pull request (anywhere is fine) to refer back to this one.  Make sure your new pull request is on a branch from latest master to minimize the chance of any possible merge conflicts.  Thanks!

**Comment by kreutpet** [comment]
recloned the repository and merged with master.
hope now everything is back on track

**Comment by watou** [comment]
I don't think this PR looks right -- there are 148 changed files!  If you do something like copy all of the files you mean to change to a temp directory, go back to your fork and do:

```shell
git checkout master
git pull
git checkout -b iec6205621-oh2
```
then copy your intended changed files back in place, and then
```shell
git commit -a -s -m "add tested IEC 62056-21 meter 1.x version to OH2"
mvn clean install
```
if it all looks and works OK, then
```shell
git push origin iec6205621-oh2
git checkout master
```
and then go to github and make a pull request.

**Comment by kreutpet** [comment]
this git is driving me nuts
on my existing fork I do have all the changes which I would like to get ride of.
github does not allow me to make a 2nd fork so I could start from scratch
above commands on my existing fork does not do anything.

so I guess I need to reset my fork

**Comment by kaikreuzer** [comment]
Thanks @kreutpet - I have done all required cleaning and created a new PR with your changes: https://github.com/openhab/openhab1-addons/pull/5230

**Pull request closed without merging**